### PR TITLE
Implement a generalized spinor basis

### PR DESCRIPTION
**Pull request by lelemmen**
**Short description**
In this PR, I will add (the beginnings of) a generalized spinor basis

**Related issues**
This PR is related to #382, but does not fix it completely.

**Don't forget:** 
- [x] if new files are created: update the Doxygen file, the collective gqcp.hpp header and the CMake files
- [x] update the documentation on [our website](https://github.com/GQCG/gqcg.github.io) and check if new Python bindings should be added

### CODE REVIEW

**Comment by lelemmen** [comment]
In order to implement `quantize()` for two-electron operators for generalized spinor bases, I need to do a small refactor. I'd like to merge this PR first, and then continue from a clean branch. @GQCG/cpp-developers Could you review this PR?

File: include/Basis/SpinorBasis/GSpinorBasis.hpp [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+     */
+    GSpinorBasis(const NuclearFramework& nuclear_framework, const std::string& basisset_name_alpha, const std::string& basisset_name_beta) :
+        GSpinorBasis(ScalarBasis<Shell>(nuclear_framework, basisset_name_alpha), ScalarBasis<Shell>(nuclear_framework, basisset_name_beta))
+    {}
+
+
+    /**
+     *  Construct a generalized spinor basis with a underlying scalar bases made by placing shells corresponding to the basisset specifications on every nucleus of the molecule
+     *
+     *  @param molecule                 the molecule containing the nuclei on which the shells should be centered
```
Comment (tmhuysen) [inline include/Basis/SpinorBasis/GSpinorBasis.hpp]:
```suggestion
     *  @param molecule                     the molecule containing the nuclei on which the shells should be centered
```

**Comment by tmhuysen** [comment]
Nice PR, approved but do address the comments before merging.

File: include/Mathematical/Representation/SquareMatrix.hpp [hunk 2]
```diff
@@ -233,14 +233,21 @@ class SquareMatrix : public MatrixX<_Scalar> {
      *  GETTERS
      */
 
-    size_t get_dim() const { return this->cols(); }  // equals this->rows()
-
+    size_t get_dim() const { return this->dimension(); }
```
[the remark this answers is not in the document]
Comment (lelemmen) [reply include/Mathematical/Representation/SquareMatrix.hpp]: It is not. There are still a couple of `get_dim()` function calls elsewhere in the library. These can be fixed when a PR addresses #330.

**Pull request merged**